Clean up the registration of recording and stimulation backends

jougs opened the pull request:
This is a rather unsorted collection of changes related to the registration of recording and stimulation backends in preparation of the module system refactoring.

med-ayssar commented on nestkernel/io_manager_impl.h:
```diff
@@ -27,28 +27,35 @@
+  }
 
+  RecordingBackendT* recording_backend = new RecordingBackendT();
+  recording_backend->pre_run_hook();
   recording_backends_.insert( std::make_pair( name, recording_backend ) );
 }
 
-template < class SBType >
+template < class StimulationBackendT >
 void
-IOManager::register_stimulation_backend( Name name )
+IOManager::register_stimulation_backend( const Name name )
 {
-  SBType* stimulation_backend = new SBType();
-  auto it = stimulation_backends_.find( name );
-  if ( it == stimulation_backends_.end() )
+  if ( stimulation_backends_.find( name ) != stimulation_backends_.end() )
   {
-    stimulation_backend->pre_run_hook();
-    stimulation_backends_.insert( std::make_pair( name, stimulation_backend ) );
+    throw BackendAlreadyRegistered( name.toString() );
   }
+
+  StimulationBackendT* stimulation_backend = new StimulationBackendT();
+  stimulation_backend->pre_run_hook();
+  stimulation_backends_.insert( std::make_pair( name, stimulation_backend ) );
 }
-}
+
+} // namespace nest
```
I would do the following:

```cpp
Class IOManager 
{
    template<class BackendT> std::map<Name, BackendT> get_backend(const BackendT& backend);
    template<class BackendT> void register_backend(const Name& name);

};

 template<> std::map<Name, StimulationBackend> IOManager::get_backend(const StimulationBackend& ){ return stimulation_backends_;}

 template<> std::map<Name, RecordingBackend> IOManager::get_backend(const RecordingBackend& ){ return recording_backends_;}
```
And then you have only one `register_backend` function that does the same for both recording and stimulation devices, depending on the type for the template class.

jougs replied:
I am also unhappy about the code duplication for the different backends, but would like to leave this for now.

My hope is that we can implement a generic `ComponentRegistry` later on, which would then support everything that can be dynamically registered (e.g. neuron and synapse models, RNG types, structural plasticity growth curves, connection rules, recording and stimulation backends).

JanVogelsang replied:
That's fine I think. As long as we add that to a kernel-team todo-list somewhere. We should start collecting all our ideas for the next hackathon following the SLI-purge one.

jougs replied:
I have captured my ideas in #2642. Please feel free to comment there in case you have further ideas.

JanVogelsang commented on nestkernel/io_manager.cpp:
```diff
@@ -327,7 +336,7 @@ IOManager::enroll_recorder( const Name& backend_name, const RecordingDevice& dev
 void
-nest::IOManager::enroll_stimulator( const Name& backend_name, StimulationDevice& device, const DictionaryDatum& params )
+nest::IOManager::enroll_stimulator( const Name backend_name, StimulationDevice& device, const DictionaryDatum& params )
 {
```
Name is just an unsigned with some functions to interpret that int, so using a reference makes no sense here. It increased the size of the parameter and adds and indirection. As the Name is not modified in either of these functions, we can safely remove the reference.

jougs commented on nestkernel/io_manager.cpp:
```diff
@@ -60,8 +61,16 @@ namespace nest
 {
-  register_recording_backends_();
-  register_stimulation_backends_();
+  register_recording_backend< RecordingBackendASCII >( "ascii" );
```
Hmmm... We could also remove the function argument altogether and take it from the instance/class inside of the function. I will look into this.

jougs replied:
Ah, the backends don't actually have a `get_ name()` function and I don't think this beautification would warrant a change of the class interface.

jougs replied:
I have written down the idea in #2642.

med-ayssar commented on nestkernel/exceptions.h:
```diff
@@ -1242,6 +1242,26 @@ class BackendNotPrepared : public KernelException
   const std::string backend_;
 };
 
+class BackendAlreadyRegistered : public KernelException
```
An `exception` namespace under `nest` namespace would be optimal!

jougs replied:
I have created #2641 to keep track of this in the context of PyNEST NG.

JanVogelsang commented:
@jougs Something went wrong there, the lines are duplicated now.

jougs commented on nestkernel/io_manager.cpp:
```diff
@@ -286,22 +295,23 @@ IOManager::cleanup()
-  std::map< Name, RecordingBackend* >::const_iterator backend;
+  auto backend = recording_backends_.find( backend_name );
+  return backend != recording_backends_.end();
   backend = recording_backends_.find( backend_name );
   return backend != recording_backends_.end();
```
```suggestion
```

jougs commented:
@JanVogelsang: I removed the duplicated lines. Also thanks for your other input.
@med-ayssar: I think I have addressed all your points. Can you please re-review? Thanks!

med-ayssar reviewed (approved):
Looks fine now!
